src/ripple/app/tx/impl/SetOracle.cpp

```diff
@@ -0,0 +1,291 @@
+            priceData.setFieldCurrency(
+                sfQuoteAsset, entry.getFieldCurrency(sfQuoteAsset));
+            pairs.emplace(tokenPairKey(entry), std::move(priceData));
+        }
+        auto const oldCount = pairs.size() > 5 ? 2 : 1;
+        // update/add/delete pairs
+        for (auto const& entry : ctx_.tx.getFieldArray(sfPriceDataSeries))
+        {
+            auto const key = tokenPairKey(entry);
+            if (!entry.isFieldPresent(sfAssetPrice))
+            {
+                // delete token pair
+                pairs.erase(key);
+            }
+            else if (auto iter = pairs.find(key); iter != pairs.end())
+            {
+                // update the price
+                iter->second.setFieldU64(
+                    sfAssetPrice, entry.getFieldU64(sfAssetPrice));
+                if (entry.isFieldPresent(sfScale))
+                    iter->second.setFieldU8(sfScale, entry.getFieldU8(sfScale));
+            }
+            else
+            {
+                // add a token pair with the price
+                STObject priceData{sfPriceData};
+                priceData.setFieldCurrency(
+                    sfBaseAsset, entry.getFieldCurrency(sfBaseAsset));
+                priceData.setFieldCurrency(
+                    sfQuoteAsset, entry.getFieldCurrency(sfQuoteAsset));
+                priceData.setFieldU64(
+                    sfAssetPrice, entry.getFieldU64(sfAssetPrice));
+                if (entry.isFieldPresent(sfScale))
+                    priceData.setFieldU8(sfScale, entry.getFieldU8(sfScale));
+                pairs.emplace(key, std::move(priceData));
+            }
+        }
+        STArray updatedSeries;
+        for (auto const& iter : pairs)
+            updatedSeries.push_back(std::move(iter.second));
+        sle->setFieldArray(sfPriceDataSeries, updatedSeries);
+        if (ctx_.tx.isFieldPresent(sfURI))
+            sle->setFieldVL(sfURI, ctx_.tx[sfURI]);
+        // LastUpdateTime is Unix time, store internally as Ripple Epoch
+        sle->setFieldU32(
+            sfLastUpdateTime, ctx_.tx[sfLastUpdateTime] - epoch_offset.count());
+
+        auto const newCount = pairs.size() > 5 ? 2 : 1;
+        auto const adjust = newCount - oldCount;
+        if (adjust != 0 && !adjustOwnerCount(ctx_, adjust))
+            return tefINTERNAL;
+
+        ctx_.view().update(sle);
+    }
+    else
+    {
+        // create
+
+        sle = std::make_shared<SLE>(oracleID);
+        sle->setAccountID(sfOwner, ctx_.tx.getAccountID(sfAccount));
+        sle->setFieldVL(sfProvider, ctx_.tx[sfProvider]);
+        if (ctx_.tx.isFieldPresent(sfURI))
+            sle->setFieldVL(sfURI, ctx_.tx[sfURI]);
+        auto const& series = ctx_.tx.getFieldArray(sfPriceDataSeries);
+        sle->setFieldArray(sfPriceDataSeries, series);
+        sle->setFieldVL(sfAssetClass, ctx_.tx[sfAssetClass]);
+        // LastUpdateTime is Unix time, store internally as Ripple Epoch
```

Comment:
Is `LastUpdateTime` in Unix time in seconds, instead of the standard milliseconds? That feels easy to mess up.

Reply:
Well, Unix time is in seconds as is Ripple Epoch. Why would it be messed up?

Reply:
Isn't Unix time usually in milliseconds?

Reply:
No, it's in seconds: https://en.wikipedia.org/wiki/Unix_time

Reply:
In JavaScript, `new Date().getTime()` returns Unix epoch time in milliseconds. In Python, `time.time()` returns units of seconds, but as a floating point that includes millisecond precision. I find this matter is generally inconsistent across languages.